File: day3.js
const foo = Bun.file("input3.txt"); // relative to cwd
// const foo = Bun.file("input3_example.txt"); // relative to cwd

foo.size; // number of bytes
foo.type; // MIME type
let rawtext = await foo.text();
// ignore line breaks
var lines = rawtext.split('\n')
const lastchar = lines[0][0]
const digits = '1234567890'
var value=''
var total = 0
console.log(lines.length)

var line_counter = 0
var next_to_char = false

function check_chars(x,y,lines){
    var out = [[1,0],[-1,0],[0,1],[0,-1],[1,1],[-1,-1],[1,-1],[-1,1]].some(d => {
        var whitelist = '1234567890.'
        try {
            // console.log(d,typeof(d[0]),typeof(d[1]),typeof(x),typeof(y))
            if (!whitelist.includes(lines[y+parseInt(d[1])][x+parseInt(d[0])])){
                // console.log('done',lines[y][x],arr,d,x,y,x+d[0],y+d[1])
                // console.log('true',arr)
                return true
            }
        } catch (err) {
            // console.error(err);
        }
         
    });
    return out
}

for (const line of lines){

    if ((value.length>0)){
        if (next_to_char){
            total+=parseInt(value)
        }else{
            console.log('not near',value)
        }
        value=''
        next_to_char=false
    }
    value=''
    next_to_char=false

    console.log(line_counter,line)
    var c=0
    for (const charpos in line){
        if (digits.includes(line[charpos])){
            value+=line[charpos]
            // check around the 8 values for a special char
            // console.log(check_chars(parseInt(charpos),line_counter,lines))
            if (check_chars(c,line_counter,lines)){
                // console.log(value,check_chars(parseInt(charpos),line_counter,lines))
                next_to_char=true
            }
        }else{
            if ((value.length>0)&next_to_char){
                if (next_to_char){
                    total+=parseInt(value)
                }else{
                    console.log('not near',value)
                }
                value=''
                next_to_char=false
            }
        }
        c+=1
    }
    line_counter+=1
}

console.log(total)
// part 1
// 548403 too low
// 549500 too low
// 551451 wrong
// 550064 from python 
// where is the fucking bug

File: day3.py
lines = open('input3.txt','r').read().split('\n')[:-1]
# lines = open('input3_example.txt','r').read().split('\n')[:-1]

print(lines)

def check(x,y):
    whitelist = '1234567890.'
    for d in [[1,0],[-1,0],[0,1],[0,-1],[1,1],[-1,-1],[1,-1],[-1,1]]:
        try:
            if lines[y+d[1]][x+d[0]] not in whitelist:
                return True
        except IndexError:
            pass
    return False

def neargear(x,y,location=False):
    for d in [[1,0],[-1,0],[0,1],[0,-1],[1,1],[-1,-1],[1,-1],[-1,1]]:
        try:
            if lines[y+d[1]][x+d[0]] =='*':
                if location:
                    return (x+d[0],y+d[1])
                else:
                    return True
        except IndexError:
            pass
    return False



total = 0 
value=''
next_to=False

for y,line in enumerate(lines):
    # check digit here. 
    if (len(value)>0) and next_to:
        total+=int(value)
    next_to=False
    value=''

    for x,char in enumerate(line):
        if char.isdigit():
            value+=char
            if check(x,y):
                next_to = True
        else:
            if (len(value)>0) and next_to:
                total+=int(value)

            next_to=False    
            value=''

print(total)   


from collections import defaultdict

gears = defaultdict(list)
# part 2 
# find all * chars, and their number pairs. 

value =''
next_to_gear =False
for y,line in enumerate(lines):
    if (len(value)>0) and near_to_gear:
        gears[g_loc].append(int(value))

    near_to_gear=False
    value=''

    for x,char in enumerate(line):
        if char.isdigit():
            value+=char
            
            if neargear(x,y):
                # print(char)
                near_to_gear=True  
                g_loc= neargear(x,y,location=True)
        else:
            if (len(value)>0) and (near_to_gear==True):
                gears[g_loc].append(int(value))

            near_to_gear=False    
            value=''

# times all the gear pairs 
part2 = 0
for k,v in gears.items():
    # print(v)
    if len(v)==2:
        print(v)
        part2+=v[0]*v[1]

print(part2)
# too low 
# 85010461